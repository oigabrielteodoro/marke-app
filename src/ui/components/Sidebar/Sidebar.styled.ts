import styled, { css } from 'styled-components'
import { AiOutlineLoading } from 'react-icons/ai'

import { theme } from 'styles'

type FileListItemProps = {
  isSelected?: boolean
  isLoading?: boolean
}

export const Container = styled.div`
  height: 100vh;
  width: 33.2rem;
  background: ${theme.colors.neutral[900]};
  display: flex;
  flex-direction: column;
  padding: 4.7rem 3.3rem;

  > img {
    margin: 0 auto;
  }
`

export const Title = styled.input`
  border: 0;
  width: 100%;
  max-width: 65%;
  white-space: nowrap;
  background: transparent;
  font-size: ${theme.font.sizes.paragraph};
`

export const NewFileButton = styled.button`
  border: 0;
  height: 4rem;
  display: flex;
  margin-top: 2.4rem;
  align-items: center;
  justify-content: center;
  border-radius: ${theme.radius.input};
  background: ${theme.colors.blue[500]};
  font-size: ${theme.font.sizes.small};
  font-weight: ${theme.font.weights.medium};
  transition: filter 0.2s;

  &:hover {
    filter: brightness(0.8);
  }

  svg {
    margin-right: 1.2rem;
  }
`

export const Separator = styled.div`
  display: flex;
  margin-top: 6rem;
  align-items: center;

  span {
    color: ${theme.colors.white};
    font-size: 1.8rem;
    margin: 0 0.5rem;
  }

  hr {
    position: relative;
    border: 0.2rem solid ${theme.colors.blue[500]};
    border-radius: 0.2rem;
    width: 100%;

    &:first-child {
      max-width: 1.3rem;
    }
  }
`

export const FileList = styled.ul`
  display: grid;
  grid-gap: 1.8rem;
  margin-top: 3.5rem;
`

export const FileListItem = styled.li<FileListItemProps>`
  opacity: 0.65;
  color: ${theme.colors.white};
  display: flex;
  align-items: center;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  transition: background 0.2s;
  height: 5rem;
  cursor: pointer;

  &:hover {
    background: rgba(255, 255, 255, 0.05);

    button {
      opacity: 1;
      visibility: visible;
    }
  }

  input {
    color: ${theme.colors.white};
  }

  button {
    background: transparent;
    height: 1.8rem;
    width: 1.8rem;
    border: 0;
    opacity: 0;
    visibility: hidden;
    transition: opacity 0.2s, visibility 0.2s, color 0.2s;
    margin-top: 0.6rem;

    &:hover {
      color: ${theme.colors.white};
    }

    & + svg {
      margin-left: 1rem;
    }
  }

  > svg {
    margin-right: 1rem;
  }

  section {
    margin-left: auto;
  }

  ${({ isLoading, isSelected }) =>
    isSelected &&
    !isLoading &&
    css`
      &::after {
        content: '';
        height: 0.8rem;
        width: 0.8rem;
        background: ${theme.colors.blue[500]};
        border-radius: 50%;
        margin-left: 1rem;
      }
    `}

  ${({ isSelected }) =>
    isSelected &&
    css`
      opacity: 1;
      background: rgba(255, 255, 255, 0.05);

      input {
        color: ${theme.colors.white};
        opacity: 1;
      }

      svg {
        color: ${theme.colors.blue[500]};
      }
    `}
`

export const SavingIcon = styled(AiOutlineLoading)`
  animation: saving 0.5s infinite;

  @keyframes saving {
    to {
      transform: rotate(0);
    }

    from {
      transform: rotate(360deg);
    }
  }
`
