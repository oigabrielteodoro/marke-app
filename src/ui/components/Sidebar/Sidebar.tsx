import React from 'react'
import { FiCheck, FiFileText, FiPlus, FiX } from 'react-icons/fi'

import { theme } from 'styles'
import { useArchive } from 'context'

import { Logo } from '..'

import * as S from './Sidebar.styled'

export function Sidebar() {
  const { archives, createArchive, inspectArchive, deleteArchive } =
    useArchive()

  return (
    <S.Container>
      <Logo />

      <S.Separator>
        <hr />
        <span>Arquivos</span>
        <hr />
      </S.Separator>

      <S.NewFileButton type='button' onClick={createArchive}>
        <FiPlus size={16} />
        Adicionar arquivo
      </S.NewFileButton>

      <S.ArchiveList>
        {archives.map((archive) => {
          const isSaved = archive.status === 'saved'
          const isEditing = archive.status === 'editing'
          const isLoading = archive.status === 'loading'

          return (
            <S.ArchiveListItem
              key={archive.id}
              isEditing={isEditing}
              isSelected={archive.active}
            >
              <FiFileText size={24} strokeWidth={1.5} />

              <a href='/' onClick={() => inspectArchive(archive.id)}>
                {archive.title}
              </a>

              <section>
                {(isSaved || isEditing) && (
                  <button
                    type='button'
                    onClick={() => deleteArchive(archive.id)}
                  >
                    <FiX size={18} color={theme.colors.white} />
                  </button>
                )}

                {isLoading && !isSaved && (
                  <S.SavingIcon
                    size={14}
                    strokeWidth={1.5}
                    color={theme.colors.blue[500]}
                  />
                )}

                {isSaved && !isEditing && (
                  <FiCheck size={16} color={theme.colors.blue[500]} />
                )}
              </section>
            </S.ArchiveListItem>
          )
        })}
      </S.ArchiveList>
    </S.Container>
  )
}
