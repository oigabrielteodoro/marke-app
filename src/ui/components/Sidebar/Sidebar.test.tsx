import React from 'react'
import { render, screen } from '@testing-library/react'

import { ArchiveProvider } from 'context'

import { Sidebar } from './Sidebar'

describe('<Sidebar />', () => {
  it('should render correctly', () => {
    const { container } = render(<Sidebar />, {
      wrapper: ({ children }) => <ArchiveProvider>{children}</ArchiveProvider>,
    })

    expect(screen.getByText('Arquivos')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
