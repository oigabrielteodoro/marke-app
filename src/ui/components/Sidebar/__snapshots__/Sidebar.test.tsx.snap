// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should render correctly 1`] = `
.c1 {
  width: 16.4rem;
  height: 4.4rem;
}

.c0 {
  height: 100vh;
  width: 33.2rem;
  background: #1e293b;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 4.7rem 3.3rem;
}

.c0 > img {
  margin: 0 auto;
}

.c3 {
  border: 0;
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 2.4rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.4rem;
  background: #1fc8e1;
  font-size: 1.4rem;
  font-weight: 500;
  -webkit-transition: filter 0.2s;
  transition: filter 0.2s;
}

.c3:hover {
  -webkit-filter: brightness(0.8);
  filter: brightness(0.8);
}

.c3 svg {
  margin-right: 1.2rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 6rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 span {
  color: #fafafa;
  font-size: 1.8rem;
  margin: 0 0.5rem;
}

.c2 hr {
  position: relative;
  border: 0.2rem solid #1fc8e1;
  border-radius: 0.2rem;
  width: 100%;
}

.c2 hr:first-child {
  max-width: 1.3rem;
}

.c4 {
  display: grid;
  grid-gap: 1.8rem;
  margin-top: 3.5rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  -webkit-transition: background 0.2s;
  transition: background 0.2s;
  color: rgba(255,255,255,0.65);
  height: 5rem;
  cursor: pointer;
  opacity: 1;
  background: rgba(255,255,255,0.05);
}

.c5 a {
  color: rgba(255,255,255,0.65);
}

.c5:hover {
  background: rgba(255,255,255,0.05);
}

.c5:hover button {
  opacity: 1;
  visibility: visible;
}

.c5 input {
  color: #fafafa;
}

.c5 button {
  background: transparent;
  height: 1.8rem;
  width: 1.8rem;
  border: 0;
  opacity: 0;
  visibility: hidden;
  -webkit-transition: opacity 0.2s,visibility 0.2s,color 0.2s;
  transition: opacity 0.2s,visibility 0.2s,color 0.2s;
  margin-top: 0.6rem;
}

.c5 button:hover {
  color: #fafafa;
}

.c5 button + svg {
  margin-left: 1rem;
}

.c5 > svg {
  margin-right: 1rem;
}

.c5 section {
  margin-left: auto;
}

.c5::after {
  content: '';
  height: 0.8rem;
  width: 0.8rem;
  background: #1fc8e1;
  border-radius: 50%;
  margin-left: 1rem;
}

.c5 a {
  color: #fafafa;
}

.c5 input {
  color: #fafafa;
  opacity: 1;
}

.c5 svg {
  color: #1fc8e1;
}

<aside
  class="c0"
>
  <img
    alt="Markee App"
    class="c1"
    src="test-file-stub"
  />
  <div
    class="c2"
  >
    <hr />
    <span>
      Arquivos
    </span>
    <hr />
  </div>
  <button
    class="c3"
    type="button"
  >
    <svg
      fill="none"
      height="16"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="12"
        x2="12"
        y1="5"
        y2="19"
      />
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
    </svg>
    Adicionar arquivo
  </button>
  <ul
    class="c4"
  >
    <li
      class="c5"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
        />
        <polyline
          points="14 2 14 8 20 8"
        />
        <line
          x1="16"
          x2="8"
          y1="13"
          y2="13"
        />
        <line
          x1="16"
          x2="8"
          y1="17"
          y2="17"
        />
        <polyline
          points="10 9 9 9 8 9"
        />
      </svg>
      <a
        href="/"
      >
        Sem t√≠tulo
      </a>
      <section>
        <button
          type="button"
        >
          <svg
            color="#fafafa"
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="color: rgb(250, 250, 250);"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
      </section>
    </li>
  </ul>
</aside>
`;
